docs/proposals/device-state.md

```diff
@@ -0,0 +1,215 @@
+states: implementable
+authors:
+    - "@JiaweiGithub"
+approvers:
+creation-date: 2024-06-02
+last-updated: 2024-06-02
+---
+
+# Support device state message.
+
+## Motivation
+The device has its own states, such as online states, offline states, etc. It is very necessary to collect these stateses and report them to the cloud.
+### Goals
+- Mapper provides a common framework code to support collecting the states of the device itself.
+- After Edgecore receives the device states message, it updates the cache database device states message and 
+  send to cloud.
+- After receiving the device states message, Cloudcore updates the device's cr states information.
+
+### Non-goals
+- Mapper only provides a general framework and does not provide device states collection code.
+
+### Device state design
+<img src="../images/device-crd/device-state.png">
+
+### Mapper design
+- In the dmi module, add a grpc interface for ReportDeviceState.
+- Call getDeviceStates based on goroutine in Mapper's DevStart.
+- GetDeviceStates regularly obtains the device states, and then sends it to edgecore through the ReportDeviceState 
+  interface.
+
+```golang
+func (gs *DeviceStates) Run() {
+	states, error := gs.Client.GetDeviceStates()
+	if error != nil {
+		klog.Errorf("GetDeviceStates failed: %v", error)
+		return
+	}
+
+	statesRequest := &dmiapi.ReportDeviceStatesRequest{
+		DeviceName:      gs.DeviceName,
+		State:           states,
+		DeviceNamespace: gs.DeviceNamespace,
+	}
+
+	log.Printf("send statesRequest", statesRequest.DeviceName, statesRequest.State)
+	if err := grpcclient.ReportDeviceStates(statesRequest); err != nil {
+		klog.Errorf("fail to report device states of %s with err: %+v", gs.DeviceName, err)
+	}
+}
+```
+
+### Edgecore design
+- In the dmiserver module, implement ReportDeviceStates and send the received device states information to TwinGroup through beehive.
+- In TwinGroup, match the "/state/update" topic, send it to the device moudle, and execute the action DeviceStateUpdate.
+- In DeviceStateUpdate, update the device states information of the sqlite database and transmit it to the cloud.
+
+```golang
+func dealDeviceStateUpdate(context *dtcontext.DTContext, resource string, msg interface{}) error {
+	message, ok := msg.(*model.Message)
+	if !ok {
+		return errors.New("msg not Message type")
+	}
+
+	updatedDevice, err := dttype.UnmarshalDeviceUpdate(message.Content.([]byte))
+	if err != nil {
+		klog.Errorf("Unmarshal device info failed, err: %#v", err)
+		return err
+	}
+	deviceID := resource
+	defer context.Unlock(deviceID)
+	context.Lock(deviceID)
+	doc, docExist := context.DeviceList.Load(deviceID)
+	if !docExist {
+		return nil
+	}
+	device, ok := doc.(*dttype.Device)
+	if !ok {
+		return nil
+	}
+
+	// state refers to definition in mappers-go/pkg/common/const.go
+	state := strings.ToLower(updatedDevice.State)
+	switch state {
+	case "online", "offline", "ok", "unknown", "disconnected":
+	default:
+		return nil
+	}
```

Comment:
Do these scenarios that return nil need to consider printing some logs?  Otherwise, the device status updatation may not be executed due to some unexpected reasons.